streaming_form_data/targets.py

```diff
@@ -35,20 +39,47 @@ def value(self):
 
 
 class FileTarget(BaseTarget):
-    def __init__(self, filename):
+    def __init__(self, filename, openmode='wb'):
         self.filename = filename
+        assert 'b' in openmode, \
+            'openmode should be binary. ' + \
+            'I.e. it should contain \'b\' character.' + \
+            'Current openmode: %s' % openmode
+        self.openmode = openmode
+        self._fd = None
 
+    def start(self):
+        self._fd = open(self.filename, self.openmode)
+
+    def data_received(self, chunk):
+        self._fd.write(chunk)
+
+    def finish(self):
+        self._fd.close()
+
+
+class FileTargetUsingRemoteName(BaseTarget):
```

Comment:
Hmm. You are right. Current code is vulnerable.
And fixing this vulnerability would be too specific to be included into the library.
I will remove this class.

But I would keep easy possibility for custom targets to get file name from HTTP. So end programmer should think about security in those case.

Reply:
Getting filenames is definitely needed. I think we're on the same page here - just make them available to the user but not use them internally.

Also, I'm somehow undecided on what's the best way to make filenames available in the `Target` interface. Thoughts?

Reply:
Hmm. What about adding an option for end user to get file name in 2 forms:

1) Original text, Lets call it something like "unsafe_raw_filename" to warn programmers additionally from using it without checks.

2) Safe and always value filename. It should have potential path truncated. Filename part should have some universal cleaning.

Cleaning of file part I would suggest to do like I did here in another project:

def clean_filename(filename):
    s = re.sub('[\\\\:\'\\[\\]/",<>&^$+*?;]', '_', filename)
    # deny special file names (Windows specific):
    s = re.sub('^(CON|PRN|AUX|NUL|COM\\d|LPT\\d)$', 'SPECIAL', s, flags=re.IGNORECASE)
    s = s if s != '' else 'empty'
    return s

I have tried to be safe for all known OSes at the same time.

It would be a good idea to give end user an option how to treat forbidden chars (for example replace them to something else OR just remove them).

Another question what to do with special names or with empty string.

Currently I don't have good beautiful idea how to update Target interface with keeping backward compatibility.

Reply:
I though again. We should not include any file name cleaning into the library.
Library should solve only one problem. And filtering is too complicated and usually specific to particular project.

Reply:
Exactly my thoughts. I would just get the filename and make it available in `Target` so the user can use it. Do you have a preference on how to change the API to do that?